core/queryalgebra/evaluation/src/test/java/org/eclipse/rdf4j/query/algebra/evaluation/function/xsd/TestDateCast.java

```diff
@@ -0,0 +1,91 @@
+/*******************************************************************************
+ * Copyright (c) 2022 Eclipse RDF4J contributors.
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Distribution License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ *******************************************************************************/
+package org.eclipse.rdf4j.query.algebra.evaluation.function.xsd;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
+import static org.eclipse.rdf4j.model.vocabulary.XSD.DATE;
+import static org.junit.Assert.assertNotNull;
+import static org.junit.Assert.fail;
+
+import org.eclipse.rdf4j.model.Literal;
+import org.eclipse.rdf4j.model.ValueFactory;
+import org.eclipse.rdf4j.model.impl.SimpleValueFactory;
+import org.eclipse.rdf4j.query.algebra.evaluation.ValueExprEvaluationException;
+import org.junit.Before;
+import org.junit.Test;
+
+public class TestDateCast {
+
+	private DateCast dateCast;
+	private final ValueFactory vf = SimpleValueFactory.getInstance();
+
+	@Before
+	public void setUp() throws Exception {
+		dateCast = new DateCast();
+	}
+
+	@Test
+	public void testCastPlainLiteral() {
+		// Arrange
+		Literal plainLit = vf.createLiteral("1999-09-09");
+		Literal result = null;
+
+		// Act
+		try {
+			result = dateCast.evaluate(vf, plainLit);
+		} catch (ValueExprEvaluationException e) {
+			fail(e.getMessage());
+		}
+
+		// Assert
+		assertNotNull(result);
+		assertThat(result.getLabel()).isEqualTo("1999-09-09");
+		assertThat(result.getDatatype()).isEqualTo(DATE);
+	}
+
+	@Test
+	public void testCastPlainLiteral_error_noDate() {
+		// Arrange
+		Literal plainLit = vf.createLiteral("1999-09-x");
+
+		// Act & Assert
+		assertThatExceptionOfType(ValueExprEvaluationException.class).isThrownBy(() -> dateCast.evaluate(vf, plainLit));
+	}
+
+	@Test
+	public void testCastPlainLiteral_error_dateTime() {
+		// Arrange
+		Literal plainLit = vf.createLiteral("1999-09-09T14:45:13.815-05:00");
+
+		// Act & Assert
+		assertThatExceptionOfType(ValueExprEvaluationException.class).isThrownBy(() -> dateCast.evaluate(vf, plainLit));
+	}
```

Comment:
I took a look at the [XSD cast table ](https://www.w3.org/TR/xpath-functions/#casting-from-primitive-to-primitive) and it looks like it should be possible to cast xsd:dateTime (and a few others) to xsd:date.

Reply:
Added the function to cast xsd:dateTime to xsd:date. I think thats all, which other types do you mean? 

<img width="594" alt="Screenshot 2022-09-16 at 21 14 16" src="https://user-images.githubusercontent.com/23375188/190714289-2fef4713-458d-44c6-a689-53b266cd1191.png">